cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py

```diff
@@ -786,7 +786,7 @@ def configure_cdk_addons():
     ceph_ep = endpoint_from_flag('ceph-storage.available')
     ceph = {}
     default_storage = ''
-    if ceph_ep:
+    if ceph_ep and ceph_ep.key():
```

Comment:
Oh mylanta.  I dislike things here :(

First -- and this isn't your fault -- I hate that we're using `ceph-storage.configured` as the name of our local charm flag.  That's *way* too easy to assume it's a relation flag coming from `ceph-storage`.  That should've been `k8s-master.ceph.configured`, but alas, it's been this way for 2yrs, so we're stuck with it.

My more pertinent beef is with `ceph_ep` out here in the wild.  You're having to do this `key` check because the handler's `@when_any` means you can't be sure which precondition brought you in here -- `k8s-master.comp.started` or `ceph-storage.configured`.

However, immediately above these lines, there's a block that does work based on `ceph-storage.configured` being set.  If `configured` is set, we know all the `ceph_ep` bits are there (or else, `configured` wouldn't be set).  Why don't we tuck this `ceph_ep` business up in there, like this:

```python
    ceph = {}
    if (is_state('ceph-storage.configured') and
            get_version('kube-apiserver') >= (1, 10)):
        cephEnabled = "true"
        # we know `available` is set because it's a precon for `ceph-storage.configured`
        ceph_ep = endpoint_from_flag('ceph-storage.available')
        b64_ceph_key = base64.b64encode(ceph_ep.key().encode('utf-8'))
        ceph['admin_key'] = b64_ceph_key.decode('ascii')
        ceph['kubernetes_key'] = b64_ceph_key.decode('ascii')
        ceph['mon_hosts'] = ceph_ep.mon_hosts()
        default_storage = hookenv.config('default-storage')
    else:
        cephEnabled = "false"
        default_storage = ''
```

I'll also eat these words if there's some reason why we need to be processing the `available` relation data outside of a `configured` guard, so lmk if that's the case.

Reply:
I understand these complaints and I agree with the first one. I should have just changed the flag name accidentally in this change. However, just knowing that ceph is configured on the master node says nothing for the data being available from the ceph cluster and therefor .key() returning a valid key value. Note the difference between ceph-storage.available and ceph-storage.configured. In the situation where I hit this issue, ceph was just allocating machines, but the relation and kubernetes were up and happy. So ceph-storage.available was true, but empty data.

Reply:
I agree, configured isn't telling us what we would like. It only is set when the key exists on <= 1.9. So any of the CSI ceph stuff that key is just meaningless. The reason for this is that originally I was going to nuke all the non-CSI ceph and just use CSI, but at the last minute I left it in for backwards compatibility. The issue there is that this flag came back and is now meaningless on CSI ceph. It shouldn't be and it should only be set if we have ceph and all things are good.

We can do one of two things:
1) Push this and then open an issue to address the ugliness of ceph-storage.configured
2) Change this to only set configured when ceph is ready with a key and then check that ceph is configured before we proceed here. That probably means moving it up as you suggested to put it under the if block for ceph-storage.configured.

If we're going to keep the <=1.9 stuff, which I think we probably should, I think 2 is the best option for us.